docs/how-to/custom-components.rst

```diff
@@ -0,0 +1,303 @@
+    >>>
+    >>> # Add subcomponents
+    >>> from icalendar import Event
+    >>> event = Event()
+    >>> event.add("uid", "123")
+    >>> comp.add_component(event)
+    >>> len(comp.subcomponents)
+    1
+
+
+Round-trip preservation
+=======================
+
+Custom components are fully preserved during round-trip parsing.
+The following example demonstrates this feature.
+
+.. code-block:: pycon
+
+    >>> from icalendar import Calendar
+    >>> original = b"""BEGIN:VCALENDAR
+    ... VERSION:2.0
+    ... PRODID:-//Test//EN
+    ... BEGIN:X-VENDOR-COMPONENT
+    ... X-VENDOR-PROP:proprietary
+    ... UID:vendor-123
+    ... END:X-VENDOR-COMPONENT
+    ... END:VCALENDAR
+    ... """
+    >>> cal = Calendar.from_ical(original)
+    >>> regenerated = cal.to_ical()
+    >>> # Custom component and its properties are preserved
+    >>> b"X-VENDOR-COMPONENT" in regenerated
+    True
+    >>> b"X-VENDOR-PROP" in regenerated
+    True
+
+
+Which ``from_ical()``?
+======================
+
+Use ``Component.from_ical()`` for standalone components and fragments.
+
+Use ``Calendar.from_ical()`` for complete iCalendar files.
+It also handles timezones.
+
+The following example shows the different usage of both methods.
+
+.. code-block:: pycon
+
+    >>> from icalendar import Component, Calendar
+    >>>
+    >>> # Standalone custom component - use Component.from_ical()
+    >>> standalone = b"""BEGIN:X-MYCOMP
+    ... UID:123
+    ... END:X-MYCOMP
+    ... """
+    >>> comp = Component.from_ical(standalone)
+    >>> comp.name
+    'X-MYCOMP'
+    >>>
+    >>> # Complete calendar - use Calendar.from_ical()
+    >>> calendar_data = b"""BEGIN:VCALENDAR
+    ... VERSION:2.0
+    ... PRODID:-//Test//EN
+    ... BEGIN:X-MYCOMP
+    ... UID:123
+    ... END:X-MYCOMP
+    ... END:VCALENDAR
+    ... """
+    >>> cal = Calendar.from_ical(calendar_data)
+    >>> len(cal.subcomponents)
+    1
+
+Create custom component subclasses
+==================================
+
+While the dynamic component creation works for most cases, you can create explicit component subclasses for custom components that need special behavior.
+The following example shows how to do so.
+
+.. code-block:: python
+
+    from icalendar import Component
+
+    class XVendorComponent(Component):
+        """Custom vendor-specific component with special behavior."""
+
+        name = "X-VENDOR"
+
+        def validate(self):
+            """Custom validation logic."""
+            required_props = ["UID", "X-VENDOR-ID"]
+            for prop in required_props:
+                if prop not in self:
+                    raise ValueError(f"Missing required property: {prop}")
+
+        def get_vendor_id(self):
+            """Convenience method for vendor ID."""
+            return self.get("X-VENDOR-ID")
+
+    # Register with the singleton factory
+    # First call to get_component_class() initializes the factory
+    Component.get_component_class("VEVENT")  # Ensure factory exists
```

Comment:
You can subclass and override this.
Then, use your component class to parse this or create your own ComponentFactory inside that.